### PR TITLE
kubelet: fix erratic restart count caused per-container restart backoff

### PR DESCRIPTION
**Pull request by yujuhong**
If a container is in the backoff period, the restart count may stop to increment.
This PR fixes that.

### CODE REVIEW

**Comment by yujuhong** [comment]
This is a new bug introduced by the pre-container restart backoff feature. It's only triggered when container is in a crashloop, where restart count would stop incrementing. Still yet to transition to writing restart counts as docker labels :(

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -476,6 +476,14 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
 		if containerStatus, found := statuses[container.Name]; found {
 			reason, ok := dm.reasonCache.Get(uid, container.Name)
 			if ok && reason == kubecontainer.ErrCrashLoopBackOff.Error() {
+				// We need to increment the restart count if we are going to
+				// move the current state to last terminated state.
+				if containerStatus.State.Terminated != nil {
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: I hope we don't increment again once we restart.

Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: We might if we don't persist it to the apiserver.

This part of the code is really messy, and we should clean it up sometime.

**Comment by samsabed** [comment]
Thanks! the restart regression around that time, hid this bug.
I'm probably missing something, Is the idea to use docker labels to persist the restart count? wouldn't it be subject to the same failure possibility (thou less likely) as persisting it to etcd via the api server.

**Comment by yujuhong** [comment]
> Thanks! the restart regression around that time, hid this bug.
> I'm probably missing something, Is the idea to use docker labels to persist the restart count? wouldn't it be subject to the same failure possibility (thou less likely) as persisting it to etcd via the api server.

kubelet doesn't persist status to apiserver as often as we'd like. The apiserver client we use has a qps limit set so that kubelet doesn't overwhelm the apiserver. If there are more updates, the situation would get worse.
On the other hand, our container GC may remove the dead container, causing incorrect counts (#13562), whether we write the count as a label or not.

**Comment by vishh** [comment]
LGTM. This is getting very tricky. Are there any plans to better organize handling restart counts?

**Comment by yujuhong** [comment]
> LGTM. This is getting very tricky. Are there any plans to better organize handling restart counts?

I think your idea of using label to store restart count is one. There are other parts that we can improve. e.g., we can filter containers using pod labels, so that we get only containers that are. We could also split the function into two parts: one reads the containers, and one fills in the reasons.

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Pull request merged**